src/catalog/dependency_manager.cpp

```diff
@@ -27,166 +236,310 @@ void DependencyManager::AddObject(CatalogTransaction transaction, CatalogEntry &
-	}
-	// add the new object to the dependents_map of each object that it depends on
-	auto &old_dependencies = dependencies_map[old_obj];
-	for (auto &dep : old_dependencies) {
-		auto &dependency = dep.get();
-		dependents_map[dependency].insert(new_obj);
-	}
+	});
 
-	// We might have to add a type dependency
-	// add the new object to the dependency manager
-	dependents_map[new_obj] = dependency_set_t();
-	dependencies_map[new_obj] = old_dependencies;
+	// Keep old dependencies
+	dependency_set_t dependents;
+	ScanDependencies(transaction, info, [&](DependencyEntry &dep) {
+		auto entry = LookupEntry(transaction, dep);
+		if (!entry) {
+			return;
+		}
+		dependents.insert(Dependency(*entry, dep.Reliant().type));
+	});
 
-	for (auto &dependency : owned_objects_to_add) {
-		dependents_map[new_obj].insert(Dependency(dependency, DependencyType::DEPENDENCY_OWNS));
-		dependents_map[dependency].insert(Dependency(new_obj, DependencyType::DEPENDENCY_OWNED_BY));
-		dependencies_map[new_obj].insert(dependency);
+	// FIXME: we should update dependencies in the future
+	// some alters could cause dependencies to change (imagine types of table columns)
+	// or DEFAULT depending on a sequence
+	if (StringUtil::CIEquals(old_obj.name, new_obj.name)) {
+		// The name was not changed, we do not need to recreate the dependency links
+		return;
 	}
-}
+	CleanupDependencies(transaction, old_obj);
 
-void DependencyManager::EraseObject(CatalogEntry &object) {
-	// obtain the writing lock
-	EraseObjectInternal(object);
-}
-
-void DependencyManager::EraseObjectInternal(CatalogEntry &object) {
-	if (dependents_map.find(object) == dependents_map.end()) {
-		// dependencies already removed
-		return;
+	for (auto &dep : dependents) {
+		auto &other = dep.entry.get();
+		DependencyInfo info {
+		    /*dependent = */ DependencyReliant {GetLookupProperties(new_obj), dep.dependency_type},
+		    /*dependency = */ DependencySubject {GetLookupProperties(other), DependencyType::DEPENDENCY_REGULAR}};
+		CreateDependency(transaction, info);
 	}
-	D_ASSERT(dependents_map.find(object) != dependents_map.end());
-	D_ASSERT(dependencies_map.find(object) != dependencies_map.end());
-	// now for each of the dependencies, erase the entries from the dependents_map
-	for (auto &dependency : dependencies_map[object]) {
-		auto entry = dependents_map.find(dependency);
-		if (entry != dependents_map.end()) {
-			D_ASSERT(entry->second.find(object) != entry->second.end());
-			entry->second.erase(object);
+
+	// For all the objects we own, re establish the dependency of the owner on the object
+	for (auto &object : owned_objects) {
+		auto &entry = object.entry.get();
+		{
+			DependencyInfo info {
+			    /*dependent = */ DependencyReliant {GetLookupProperties(new_obj), DependencyType::DEPENDENCY_OWNED_BY},
+			    /*dependency = */ DependencySubject {GetLookupProperties(entry), DependencyType::DEPENDENCY_REGULAR}};
+			CreateDependency(transaction, info);
+		}
+		{
+			DependencyInfo info {
```

Comment:
I'll have a look at combining this information in a single dependent<->dependency instead of two.